src/env-inl.h

```diff
@@ -327,16 +327,48 @@ inline Environment* Environment::GetCurrent(
   return GetFromCallbackData(info.Data());
 }
 
-inline Environment* Environment::GetFromCallbackData(v8::Local<v8::Value> val) {
+Environment* Environment::GetFromCallbackData(v8::Local<v8::Value> val) {
   DCHECK(val->IsObject());
   v8::Local<v8::Object> obj = val.As<v8::Object>();
-  DCHECK_GE(obj->InternalFieldCount(), 1);
-  Environment* env =
-      static_cast<Environment*>(obj->GetAlignedPointerFromInternalField(0));
+  DCHECK_GE(obj->InternalFieldCount(),
+            BaseObject::kInternalFieldCount);
+  Environment* env = Unwrap<BaseObject>(obj)->env();
   DCHECK(env->as_callback_data_template()->HasInstance(obj));
   return env;
 }
 
+template <typename T>
+Environment::BindingScope<T>::BindingScope(Environment* env) : env(env) {
+  v8::Local<v8::Object> callback_data;
+  if (!env->MakeBindingCallbackData<T>().ToLocal(&callback_data))
+    return;
+  data = Unwrap<T>(callback_data);
+
+  // No nesting allowed currently.
+  CHECK_EQ(env->current_callback_data(), env->as_callback_data());
+  env->set_current_callback_data(callback_data);
+}
+
+template <typename T>
+Environment::BindingScope<T>::~BindingScope() {
+  env->set_current_callback_data(env->as_callback_data());
+}
+
+template <typename T>
+v8::MaybeLocal<v8::Object> Environment::MakeBindingCallbackData() {
+  v8::Local<v8::Function> ctor;
+  v8::Local<v8::Object> obj;
+  if (!as_callback_data_template()->GetFunction(context()).ToLocal(&ctor) ||
+      !ctor->NewInstance(context()).ToLocal(&obj)) {
```

Comment:
Sorry for noticing this late - but I came across this change after a recent rebase, and I think this constraint about `BaseObject` is incorrect (see https://github.com/nodejs/node/pull/26382#issuecomment-519081673). Context-independent templates should not point to context-dependent objects, but now every templates created by `Environment::FunctionTemplate`, which are supposed to be context-independent, point to  different callback data that are context-dependent `v8::Object`s that point to `BaseObject`s, making the templates non-snapshottable (and theoratically just incorrect, but this is only validated by V8's serializer and not at runtime). This was already violated by https://github.com/nodejs/node/pull/26382 and this PR just splits one context-dependent callback data into multiple context-dependent ones.

Reply:
Note that even before https://github.com/nodejs/node/pull/26382, when we were using `v8::External`, it was still somewhat limited - by obtaining the data directly through whatever passed into `FunctiontTemplate::New`, we'll always end up with the same set of callback data in the function template callback even if we switch the invoking context. This PR only makes the binding data variable by context, but it ties the FunctionTemplates to the set of the data created by the main context. If the intention is to make the binding data variable by the invoking context, we need to associate the binding data with the contexts and obtain them from there (e.g. through embedder slots, or private properties on the global object, something like `GetBindingData<T>(args.GetIsolate()->GetCurrentContext())`), instead of associating them with and and obtaining them from the function template callback data (`Unwrap<T>(args.Data())` currently being done here), since there's always only one set of data per function template.

Reply:
@joyeecheung Yeah, I know about these issue, but this PR doesn’t really seem to make anything worse?

There is a commit in https://github.com/nodejs/node/pull/32761 that would switch this back to using `External`s for snapshotting, if we don’t end up with a better solution.



> If the intention is to make the binding data variable by the invoking context, (for example, so that different vm contexts have different set of builtins), we need to associate the binding data with the contexts and obtain them from there (e.g. through embedder slots, or private properties on the global object, something like `GetBindingData<T>(args.GetIsolate()->GetCurrentContext())`), instead of associating them with and and obtaining them from the function template callback data (`Unwrap<T>(args.Data())` currently being done here), since there's always only one set of data per function template.

Err, yeah, we could do that, but if you look it that way we’re just using `FunctionTemplate`s and `ObjectTemplate`s in a wrong way to begin with. It would be nicer if we could just create context-dependent templates, I would say.

Reply:
… or, even better, if we could provide data to the per-context instantiations of the templates, separately.

Reply:
> Yeah, I know about these issues, but this PR doesn’t really seem to make anything worse?

I think this makes it worse in that it now forces the FunctionTemplates to be associated with context-dependent `BaseObject`s with the `BindingData` machinery. In addition `BaseObject` implicitly implies that the object is tied to the main context. If we were to add an indirection, we could add an indirection to `v8::External` or a different abstraction based on that, but an indirection based on `BaseObject` still doesn't make too much sense to me.

> It would be nicer if we could just create context-dependent templates, I would say.

From previous discussions I think that was discouraged and it is preferred to obtain the properties from contexts. `FunctionTemplate`s and `ObjectTemplates` essentially belong to the isolate while the `Function` and `Object` instantiated from them belong to the context.

Reply:
> > Yeah, I know about these issues, but this PR doesn’t really seem to make anything worse?
> 
> I think this makes it worse in that it now forces the FunctionTemplates to be associated with context-dependent `BaseObject`s with the `BindingData` machinery.

… but previously it was also associated with a context-dependent object?

> In addition `BaseObject` implicitly implies that the object is tied to the main context

In what way? We can and do create `BaseObject`s in core that are associated with Contexts coming from `vm.createContext()`.

> If we were to add an indirection, we could add an indirection based on `v8::External` directly, but an indirection based on `BaseObject`(which in turn relies on `Object`) still doesn't make too much sense to me.

I’m not sure what you have in mind, but as said above, #32761 does contain a commit that would switch this back to `v8::External` for snapshotting.

> > It would be nicer if we could just create context-dependent templates, I would say.
> 
> From previous discussions I think that was discouraged and it is preferred to obtain the properties from contexts.

Yeah, hence the addition that it would be even better for the data associated with the template to be set or filled per-context.

> `FunctionTemplate`s and `ObjectTemplates` essentially belong to the isolate while the `Function` and `Object` instantiated from them belong to the context.

Yeah, I know. I don’t think we’re disagreeing here.